src/Persistent_cohomology/include/gudhi/Persistence_on_a_line.h

```diff
@@ -0,0 +1,148 @@
+/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+ *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+ *    Author(s): Marc Glisse
+ *
+ *    Copyright (C) 2022 Inria
+ *
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+#ifndef PERSISTENCE_ON_A_LINE_H_
+#define PERSISTENCE_ON_A_LINE_H_
+
+#include <vector>
+#include <limits>
+#include <type_traits>
+#include <stdexcept>
+#include <gudhi/Debug_utils.h>
+
+namespace Gudhi::persistent_cohomology {
+/**
+ * \brief Computes the persistent homology of the sublevelsets of a PL function defined on ℝ in linear time.
+ *
+ * \ingroup persistent_cohomology
+ *
+ * Not all input values appear in the output (they would be part of pairs of length 0
+ * with a simplicial / cubical complex).
+ *
+ * @param[in] input Range of filtration values.
+ * @param[out] out Functor that is called as `out(birth, death)` for each persistence pair.
+ * By convention, it is also called one last time on the minimum and `std::numeric_limits<Filtration>::infinity()`.
+ * @param[in] lt Functor that compares 2 filtration values.
+ *
+ * \author Marc Glisse
+ */
+template<class FiltrationRange, class OutputFunctor, class Compare = std::less<>>
+void compute_persistence_of_function_on_line(FiltrationRange const& input, OutputFunctor&& out, Compare&& lt = {}) {
+  using std::begin;
+  using std::end;
+  auto it = begin(input);
+  auto stop = end(input);
+  if (it == stop) return;
+  typedef std::decay_t<decltype(*it)> Filtration;
+  std::vector<Filtration> data; // Invariant: data contains a sequence of type 1 9 2 8 3 7 ...
+  Filtration v;
+  auto le = [&lt](auto& x, auto& y){ return !lt(y, x); };
+  auto ge = [&lt](auto& x, auto& y){ return !lt(x, y); };
+  auto gt = [&lt](auto& x, auto& y){ return  lt(y, x); };
+  data.push_back(*it++);
+state1:
```

Comment:
I don't know if it would help to understand the state machine, by commenting the state in the code (like it is done in the publication) like this:
```cpp
// state1 : input[i] >= input[i+1]
state1:
...
// state12 :  input[i] < input[i+1]
state12:
...
// state132 :  input[i] < input[i+2] < input[i+1]
state132:
...
// state132 :  input[i+1] < input[i+2] < input[i]
state312:
...
```
I mention it because it perturbs me at first code reading.

Reply:
Yes, thanks !